libsql-server/src/replication/snapshot.rs

```diff
@@ -99,57 +102,129 @@ pub async fn find_snapshot_file(
     pub fn new(db_path: &Path, log_id: Uuid, callback: SnapshotCallback) -> anyhow::Result<Self> {
-        // we create a 0 sized channel, in order to create backpressure when we can't
-        // keep up with snapshop creation: if there isn't any ongoind comptaction task processing,
-        // the compact does not block, and the log is compacted in the background. Otherwise, the
-        // block until there is a free slot to perform compaction.
-        let (sender, mut receiver) = mpsc::channel::<(LogFile, PathBuf, u32)>(1);
+        // a directory containing logs that need compaction
+        let compact_queue_dir = db_path.join("to_compact");
+        std::fs::create_dir_all(&compact_queue_dir)?;
+        let (sender, mut receiver) = mpsc::channel::<(LogFile, PathBuf)>(8);
         let mut merger = SnapshotMerger::new(db_path, log_id)?;
-        let db_path = db_path.to_path_buf();
         let snapshot_dir_path = snapshot_dir_path(&db_path);
-        let _handle = tokio::task::spawn(async move {
-            while let Some((file, log_path, size_after)) = receiver.recv().await {
-                match perform_compaction(&db_path, file, log_id).await {
-                    Ok((snapshot_name, snapshot_frame_count)) => {
-                        tracing::info!("snapshot `{snapshot_name}` successfully created");
-
-                        let snapshot_file = snapshot_dir_path.join(&snapshot_name);
-                        if let Err(e) = (*callback)(&snapshot_file) {
-                            tracing::error!("failed to call snapshot callback: {e}");
-                            break;
-                        }
-
-                        if let Err(e) = merger
-                            .register_snapshot(snapshot_name, snapshot_frame_count, size_after)
-                            .await
-                        {
-                            tracing::error!(
-                                "failed to register snapshot with snapshot merger: {e}"
-                            );
-                            break;
-                        }
-
-                        if let Err(e) = std::fs::remove_file(&log_path) {
-                            tracing::error!(
-                                "failed to remove old log file `{}`: {e}",
-                                log_path.display()
-                            );
-                            break;
-                        }
-                    }
-                    Err(e) => {
-                        tracing::error!("fatal error creating snapshot: {e}");
-                        break;
-                    }
+
+        let db_path = db_path.to_path_buf();
+        // We gather pending snapshots here, so new snapshots don't interfere.
+        let pending = pending_snapshots_list(&compact_queue_dir)?;
+        // FIXME(marin): we somehow need to make this code more robust. How to deal with a
+        // compaction error?
+        tokio::task::spawn(async move {
```

Comment:
Since we don't join on this task it can fail and the database will continue to run? Is this a problem? Should we be waiting for this work to complete before we return from this function?

Reply:
This is a background job, if it exist the receiver gets dropped, and we can' send compact jobs anymore, we will know the loop exited